### DISCUSSION TITLE
Robot spins continuous with command of right Xbox controller

### DISCUSSION THREAD

**Discussion by kreidljj** [BroncBotz3481/YAGSL-Example#81] [Q&A] [answered]
Overall implementation is going great - thank you for this!  We use Xbox controllers, the left stick is properly working commanding forward and back, left and right.  I can't seem to get the right stick (axis 4) to control rotation.  The robot will spin continuously in the direction it is commanded, until the right stick is moved in the y axis, it will then eventually stop rotating.  I started from the example project with minor updates to cover our modules and the Xbox controller.  I know it is something simple, any ideas?  Below are snips of the code that I 'think' are pertinent.  Thanks in advance - Jason

  XboxController driverXbox = new XboxController(0);
  AbsoluteDrive closedAbsoluteDrive = new AbsoluteDrive(drivebase,
  () -> MathUtil.applyDeadband(driverXbox.getLeftY(),
                               OperatorConstants.LEFT_Y_DEADBAND),
  () -> MathUtil.applyDeadband(driverXbox.getLeftX(),
                               OperatorConstants.LEFT_X_DEADBAND),
  () -> MathUtil.applyDeadband(driverXbox.getRightX(),
                               OperatorConstants.LEFT_X_DEADBAND),
  () -> MathUtil.applyDeadband(driverXbox.getRightY(),
                               OperatorConstants.LEFT_Y_DEADBAND),                                                          
  false);

    AbsoluteFieldDrive closedFieldAbsoluteDrive = new AbsoluteFieldDrive(drivebase,
     () ->
         MathUtil.applyDeadband(driverXbox.getLeftY(),
                                OperatorConstants.LEFT_Y_DEADBAND),
     () -> MathUtil.applyDeadband(driverXbox.getLeftX(),
                                  OperatorConstants.LEFT_X_DEADBAND),
     () -> MathUtil.applyDeadband(driverXbox.getRightX(),
                                  OperatorConstants.LEFT_X_DEADBAND),
                                  false);

    TeleopDrive simClosedFieldRel = new TeleopDrive(drivebase,
    () -> MathUtil.applyDeadband(driverXbox.getLeftY(),
                                 OperatorConstants.LEFT_Y_DEADBAND),
    () -> MathUtil.applyDeadband(driverXbox.getLeftX(),
                                 OperatorConstants.LEFT_X_DEADBAND),
    () -> MathUtil.applyDeadband(driverXbox.getRightX(),
                                 OperatorConstants.LEFT_X_DEADBAND),
    () -> true, false, false); //headingCorrection was true

    TeleopDrive closedFieldRel = new TeleopDrive(drivebase,
      () -> MathUtil.applyDeadband(driverXbox.getLeftY(),
      OperatorConstants.LEFT_Y_DEADBAND),
      () -> MathUtil.applyDeadband(driverXbox.getLeftX(),
      OperatorConstants.LEFT_X_DEADBAND),
      () -> MathUtil.applyDeadband(driverXbox.getRightX(),
      OperatorConstants.LEFT_X_DEADBAND),
() -> true, false, false); //headingCorrection was true

**Comment by thenetworkgrinch** [comment]
Have you tried inverting your gyro? I have a setting for it in one of the configs, also turn off heading correction. Maybe try turning off heading correction first.

**Reply by kreidljj** [reply]
Thank you for the prompt reply!  I turned off heading correction before, that didn't seem to solve it.  Just tried switching invertedIMU from true to false with no luck.  Should I try switching heading correction back on with the gyro switched? - Jason

**Reply by thenetworkgrinch** [reply]
Can you post a link to your code? 

Have you tried this in simulation too? Is it the same result?

Are you accidentally not using `SwerveController`?

Is your PID for `swervecontroller.json` negated in P?

When you rotate your robot are you counter-clockwise-positive?

**Reply by kreidljj** [reply]
Below is a link to the code (I hope I am doing that right!) - let me know if you're unable to get to it.  I started with the YAGSL example project and only made adjustments as needed for our geometry / components.  In simulation it doesn't behave exactly the same way, the robot will rotate in the direction the right stick is pushed, then rotate back to the original heading when the stick is released.  - Jason
https://github.com/RedRaiderRobotics6381/MiniBot-YAGSL-Example-V0.git

**Reply by thenetworkgrinch** [reply]
Try removing the limitVelocity function and using the direct result of getTargetSpeeds. You may have forgotten to set the chassis weight to use this function.

https://github.com/RedRaiderRobotics6381/MiniBot-YAGSL-Example-V0/blob/main/src/main/java/frc/robot/commands/swervedrive/drivebase/AbsoluteDrive.java#L79-L85

**Reply by kreidljj** [reply]
I did not change the chassis weight, that is fixed now.  I tried removing limitVelocity and directly passing the getTargetSpeeds with no change.  I did load a fresh YAGSL example and found that while using the keyboard instead of the Xbox controller in simulation I see the desired behavior with axis 2.  The difference being that the keyboard counts up from 0 to 1 and does not return to zero when you release the key like the Xbox controller does.  This again points to something simple that I am overlooking.  - Jason

**Reply by thenetworkgrinch** [reply]
Have you looked at it via frc-web-components app?

https://github.com/frc-web-components/app/releases/tag/app-v0.4.2

If the getTargetSpeeds function is failing it could be either the pid constants or a bad imy/gyro config.

**Reply by kreidljj** [reply]
I got it working - kind of...
In RobotContainer I changed:
drivebase.setDefaultCommand(!RobotBase.isSimulation() ? closedFieldRel : simClosedFieldRel);
was:
drivebase.setDefaultCommand(!RobotBase.isSimulation() ? closedAbsoluteDrive : closedFieldAbsoluteDrive);
I also set headingCorrection to False

Now for the "kind of" part of fixed.  The changes above solved the robot spinning uncontrollably.  I am unable to get it to operate in field relative drive mode.  I tried a Pegion2 IMU, I tried inverting the IMU, I adjusted the robot mass.  It behaves correctly in simulation.  Any more ideas?  Thanks a ton! - Jason

**Reply by thenetworkgrinch** [reply]
Can you send a video of your controller + the frc-web-components app with the swerve drive widget?

**Reply by kreidljj** [reply]
I will, the robot isn't with me until next week.  The FRC-Web-Components is very slick!  On chiefdelphi one of the intro discussions says that YAGSL builds a SmartDashboard that shows the modules and more information.  How are these dashboards loaded, I can't seem to find them?  - Jason

**Reply by thenetworkgrinch** [reply]
I got rid of them bc it kept using too many resources on my computer and would crash others. Thats why i switch YAGSL to use frc-web-components

**Reply by kreidljj** [reply, marked as answer]
I would like to document our error and close this question out.  First, there was / is nothing wrong with YAGSL - it works as designed.  The robot that I was implementing this on is a very compact swerve robot, like 10.5" center to center wheel distance compact.  This meant that the RoboRio was mounted at a 45 degree angle (in the Z) to the chassis, with a NavX2 mounted on it.  This twist in the gyro relative to the frame, even with a full calibration was a problem.  The solution was to mount a Pigeon2 to the robot, but that also meant adding a CANivore, because the can bus became saturated.  Getting all of the firmware in the right version was a challenge, just because there are some newer 2024 firmware's starting to trickle out, sometimes newer isn't better (yet).

I just want to say thank you again for helping me through this, your request to video the dashboard is what tipped me off that the gyro data wasn't coming over. - Jason
